libs/vkd3d/state.c

```diff
@@ -5522,29 +5522,72 @@ static HRESULT d3d12_pipeline_create_private_root_signature(struct d3d12_device
         VkPipelineBindPoint bind_point, const struct d3d12_pipeline_state_desc *desc,
         struct d3d12_root_signature **root_signature)
 {
-    const struct D3D12_SHADER_BYTECODE *bytecode;
-    ID3D12RootSignature *object = NULL;
-    HRESULT hr;
+    const struct D3D12_SHADER_BYTECODE *bytecodes[VKD3D_MAX_SHADER_STAGES];
+    unsigned int num_bytecodes = 0;
+    unsigned int i;
+
+    /* Any shader stage can define the private root signature.
+     * If multiple stages emit it, they have to match or runtime complains. */
 
     if (bind_point == VK_PIPELINE_BIND_POINT_COMPUTE)
-        bytecode = &desc->cs;
+    {
+        bytecodes[num_bytecodes++] = &desc->cs;
+    }
     else if (desc->ms.BytecodeLength)
-        bytecode = &desc->ms;
+    {
+        bytecodes[num_bytecodes++] = &desc->as;
+        bytecodes[num_bytecodes++] = &desc->ms;
+        bytecodes[num_bytecodes++] = &desc->ps;
+    }
     else
-        bytecode = &desc->vs;
+    {
+        bytecodes[num_bytecodes++] = &desc->vs;
+        bytecodes[num_bytecodes++] = &desc->hs;
+        bytecodes[num_bytecodes++] = &desc->ds;
+        bytecodes[num_bytecodes++] = &desc->gs;
+        bytecodes[num_bytecodes++] = &desc->ps;
+    }
 
-    if (!bytecode->BytecodeLength)
-        return E_INVALIDARG;
+    *root_signature = NULL;
 
-    if (FAILED(hr = ID3D12Device12_CreateRootSignature(&device->ID3D12Device_iface, 0,
-            bytecode->pShaderBytecode, bytecode->BytecodeLength,
-            &IID_ID3D12RootSignature, (void**)&object)))
-        return hr;
+    for (i = 0; i < num_bytecodes; i++)
+    {
+        const struct D3D12_SHADER_BYTECODE *bytecode = bytecodes[i];
+        ID3D12RootSignature *object = NULL;
 
-    *root_signature = impl_from_ID3D12RootSignature(object);
-    d3d12_root_signature_inc_ref(*root_signature);
-    ID3D12RootSignature_Release(object);
-    return hr;
+        if (!bytecode->BytecodeLength)
+            continue;
+
+        if (SUCCEEDED(ID3D12Device12_CreateRootSignature(&device->ID3D12Device_iface, 0,
```

Comment:
does an error return here imply "shader doesn't define root sig"?

Reply:
ye, it fails if there is no root signature. I could guard it with a check for "is there even any RS in here" to make it clearer.

Reply:
Fixed.